compiler/passes/src/flattening/flattener.rs

```diff
@@ -95,30 +121,65 @@ impl<'a> Flattener<'a> {
         core::mem::take(&mut self.returns)
     }
 
-    /// Constructs a guard from the current state of the condition stack.
-    pub(crate) fn construct_guard(&mut self) -> Option<Expression> {
-        match self.condition_stack.is_empty() {
-            true => None,
-            false => {
-                let (first, rest) = self.condition_stack.split_first().unwrap();
-                Some(rest.iter().cloned().fold(first.clone(), |acc, condition| {
-                    // Construct the binary expression.
-                    Expression::Binary(BinaryExpression {
-                        op: BinaryOperation::And,
-                        left: Box::new(acc),
-                        right: Box::new(condition),
-                        span: Default::default(),
-                        id: {
-                            // Create a new node ID for the binary expression.
-                            let id = self.node_builder.next_id();
-                            // Set the type of the node ID.
-                            self.type_table.insert(id, Type::Boolean);
-                            id
-                        },
-                    })
-                }))
+    /// Construct a guard from the current state of the condition stack.
+    ///
+    /// That is, a boolean expression which is true iff we've followed the branches
+    /// that led to the current point in the Leo code.
+    pub(crate) fn construct_guard(&mut self) -> Option<(Expression, Vec<Statement>)> {
+        if self.condition_stack.is_empty() {
+            return None;
+        }
+
+        // All guards up to a certain point in the stack should be constructed.
+        // Find the first unconstructed one. Start the search at the end so we
+        // don't repeatedly traverse the whole stack with repeated calls to this
+        // function.
+        let start_i = (0..self.condition_stack.len())
+            .rev()
+            .take_while(|&i| matches!(self.condition_stack[i], Guard::Unconstructed(_)))
+            .last()
+            .unwrap_or(self.condition_stack.len());
+
+        let mut statements = Vec::with_capacity(self.condition_stack.len() - start_i);
+
+        for i in start_i..self.condition_stack.len() {
+            let expression = self.condition_stack[i].expression();
+            if i == 0 {
+                self.condition_stack[0] = Guard::Constructed(expression);
```

Comment:
Is there a way this could be a non-trivial expression (not an `Identifier`) and if so, should we create an assignment statement for it?